File: apps/android/app/src/test/java/dev/tohure/android_rust_test/ui/transfer/TransferViewModelTest.kt
package dev.tohure.android_rust_test.ui.transfer

import kotlinx.coroutines.ExperimentalCoroutinesApi
import dev.tohure.android_rust_test.adapter.ContractMessages
import dev.tohure.android_rust_test.adapter.FakeCoreFinanciero
import dev.tohure.android_rust_test.contract.ContractSource
import dev.tohure.android_rust_test.contract.MessageSource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.test.setMain
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import uniffi.core_financiero.Account
import uniffi.core_financiero.DomainException
import uniffi.core_financiero.TransferResult

private class FakeContract : ContractSource {
    override fun initialAccounts() = listOf(
        Account("00219100123456789047", "Ana Quispe", "5000.00"),
        Account("01122000987654321065", "Luis Ramos", "1200.50"),
    )

    override fun demoKeyHex() = "00010203"
    override fun demoNonceHex() = "0001"
}

private class FakeMessages(private val map: Map<String, String>) : MessageSource {
    override fun messages() = map
}

@OptIn(ExperimentalCoroutinesApi::class)
class TransferViewModelTest {
    private val testDispatcher = StandardTestDispatcher()

    @Before
    fun setup() {
        Dispatchers.setMain(testDispatcher)
    }

    @After
    fun tearDown() {
        Dispatchers.resetMain()
    }

    private fun messages() = ContractMessages(
        FakeMessages(
            mapOf(
                "MismaCuenta" to "La cuenta de origen y la de destino son la misma.",
                "SaldoInsuficiente" to "Saldo insuficiente: tenés {available} y se necesitan {required}.",
            ),
        ),
    )

    @Test
    fun theInitialAccountsComeFromTheContractNotFromCode() {
        val vm = TransferViewModel(FakeCoreFinanciero(), FakeContract(), messages())
        assertEquals(2, vm.uiState.value.accounts.size)
        assertEquals("Ana Quispe", vm.uiState.value.accounts[0].holder)
    }

    @Test
    fun theAmountFieldRejectsMoreThanTwoDecimals() {
        val vm = TransferViewModel(FakeCoreFinanciero(), FakeContract(), messages())
        vm.amountChanged("100.12")
        assertEquals("100.12", vm.uiState.value.amount)
        // Filtro de TEXTO, no regla de negocio: el core igual rechaza (tr-007), pero el
        // usuario no tiene que llegar hasta ahí.
        vm.amountChanged("100.123")
        assertEquals("100.12", vm.uiState.value.amount)
    }

    @Test
    fun aSuccessfulTransferLandsInTheStateAfterTheSimulatedLatency() = runTest {
        val core = FakeCoreFinanciero(
            nextTransfer = Result.success(
                TransferResult(
                    accounts = listOf(
                        Account("00219100123456789047", "Ana Quispe", "4899.99"),
                        Account("01122000987654321065", "Luis Ramos", "3100.00"),
                    ),
                    itfFee = "0.01",
                    totalDebited = "100.01",
                    receipt = "TRF-0001",
                    simulatedLatencyMs = 120u,
                ),
            ),
        )
        val vm = TransferViewModel(core, FakeContract(), messages())
        vm.transfer()
        advanceUntilIdle()
        assertEquals("0.01", vm.uiState.value.result?.itfFee)
        assertEquals("4899.99", vm.uiState.value.accounts[0].balance)
        assertFalse(vm.uiState.value.isLoading)
        assertNull(vm.uiState.value.error)
    }

    @Test
    fun aDomainErrorBecomesUserTextAndTurnsOffTheSpinner() = runTest {
        val core = FakeCoreFinanciero()
        core.failNextTransfer(DomainException.InsufficientFunds("50.00", "100.01"))
        val vm = TransferViewModel(core, FakeContract(), messages())
        vm.transfer()
        advanceUntilIdle()
        assertEquals(
            "Saldo insuficiente: tenés 50.00 y se necesitan 100.01.",
            vm.uiState.value.error,
        )
        // El bug clásico: el catch se olvida de apagar el spinner y la pantalla queda
        // cargando para siempre.
        assertFalse(vm.uiState.value.isLoading)
    }

    @Test
    fun aFailedTransferClearsThePreviousResult() = runTest {
        val core = FakeCoreFinanciero()
        val vm = TransferViewModel(core, FakeContract(), messages())
        vm.transfer()
        advanceUntilIdle()
        assertNotNull("precondición: la primera transferencia dejó un resultado", vm.uiState.value.result)

        // Un fallo no puede dejar el comprobante anterior en pantalla debajo del error: parece
        // que la segunda transferencia surtió efecto parcial.
        core.failNextTransfer(DomainException.InsufficientFunds("50.00", "100.01"))
        vm.transfer()
        advanceUntilIdle()
        assertNull(vm.uiState.value.result)
        assertNotNull(vm.uiState.value.error)
    }

    @Test
    fun clearErrorConsumesTheError() = runTest {
        val core = FakeCoreFinanciero()
        core.failNextTransfer(DomainException.SameAccount())
        val vm = TransferViewModel(core, FakeContract(), messages())
        vm.transfer()
        advanceUntilIdle()
        vm.clearError()
        assertNull(vm.uiState.value.error)
    }
}

File: apps/ios/ios-rust-testTests/TransferViewModelTest.swift
import Foundation
import Testing

@testable import ios_rust_test

@MainActor
@Suite("ViewModel de Transferencia")
struct TransferViewModelTest {
    private func makeViewModel(_ core: FakeCoreFinanciero = FakeCoreFinanciero()) throws
        -> TransferViewModel
    {
        let bundle = Bundle(for: BundleToken.self)
        return TransferViewModel(
            core: core,
            contract: try BundleContractSource(bundle: bundle),
            messages: ContractMessages(source: try BundleMessageSource(bundle: bundle))
        )
    }

    @Test("arranca con las dos cuentas del contrato, origen y destino ya elegidos")
    func startsWithTheContractAccounts() throws {
        let vm = try makeViewModel()
        #expect(vm.state.accounts.count == 2)
        #expect(vm.state.origin == "00219100123456789047")
        #expect(vm.state.destination == "01122000987654321065")
        #expect(vm.state.amount.isEmpty)
        #expect(vm.state.result == nil)
    }

    @Test("el campo de monto acepta como máximo 2 decimales")
    func theAmountFieldAcceptsAtMostTwoDecimals() throws {
        let vm = try makeViewModel()
        vm.amountChanged("100")
        #expect(vm.state.amount == "100")
        vm.amountChanged("100.5")
        #expect(vm.state.amount == "100.5")
        vm.amountChanged("100.50")
        #expect(vm.state.amount == "100.50")
        // El tercer decimal NO entra. Es un filtro de TEXTO, no una validación: quien
        // rechaza el monto sigue siendo el core, y tr-007 lo prueba en el test de contrato.
        vm.amountChanged("100.501")
        #expect(vm.state.amount == "100.50")
        // Tampoco entra una coma: el .decimalPad con locale es_PE la ofrece.
        vm.amountChanged("100,50")
        #expect(vm.state.amount == "100.50")
    }

    @Test("una transferencia válida guarda el resultado y actualiza los saldos")
    func aValidTransferUpdatesTheBalances() async throws {
        let core = FakeCoreFinanciero()
        core.transferResult = { _, _ in
            TransferResult(
                accounts: [
                    Account(id: "00219100123456789047", holder: "Ana Quispe", balance: "4899.99"),
                    Account(id: "01122000987654321065", holder: "Luis Ramos", balance: "1300.50"),
                ],
                itfFee: "0.01",
                totalDebited: "100.01",
                receipt: "TRF-9047-1065-10000",
                simulatedLatencyMs: 0
            )
        }
        let vm = try makeViewModel(core)
        vm.amountChanged("100.00")
        await vm.transfer()
        #expect(vm.state.result?.receipt == "TRF-9047-1065-10000")
        #expect(vm.state.result?.itfFee == "0.01")
        #expect(vm.state.accounts[0].balance == "4899.99")
        #expect(vm.state.isLoading == false)
        #expect(vm.state.error == nil)
    }

    @Test("un error apaga el spinner y guarda el texto de usuario")
    func anErrorTurnsOffTheSpinner() async throws {
        let core = FakeCoreFinanciero()
        core.transferResult = { _, _ in
            throw DomainError.InsufficientFunds(available: "1200.50", required: "10000.50")
        }
        let vm = try makeViewModel(core)
        vm.amountChanged("10000.00")
        await vm.transfer()
        // El bug clásico es el catch que se olvida de apagar el spinner.
        #expect(vm.state.isLoading == false)
        #expect(vm.state.result == nil)
        #expect(
            vm.state.error == "Saldo insuficiente: tienes 1200.50 y se necesitan 10000.50."
        )
    }

    @Test("clearError existe y el error se consume")
    func clearErrorConsumesTheError() async throws {
        let core = FakeCoreFinanciero()
        core.transferResult = { _, _ in throw DomainError.SameAccount }
        let vm = try makeViewModel(core)
        await vm.transfer()
        #expect(vm.state.error != nil)
        vm.clearError()
        #expect(vm.state.error == nil)
    }

    private final class BundleToken {}
}
